fix(api): Correct error mapping to use HTTP status codes

Mostafa-khaled21 opened the pull request:
The previous implementation was mapping the API's custom message code instead of the standard HTTP status code. This resulted in many errors being incorrectly categorized as 'UnknownException'.

Mostafa-khaled21 committed:
fix(api): Correct error mapping to use HTTP status codes

The previous implementation was mapping the API's custom message code instead of the standard HTTP status code. This resulted in many errors being incorrectly categorized as 'UnknownException'.

poula22 commented on repository/media/src/main/java/com/giraffe/media/utils/SafeCall.kt:
```diff
@@ -43,14 +43,14 @@ object SafeCall {
-            2, 9, 43, 46 -> NetworkException()
-            11, 15, 44 -> ServerErrorException()
-            24 -> TimeoutException()
+            401  -> UnauthorizedException()
+            405  -> InvalidRequestMethodException()
+            403, 429 -> AccessDeniedException()
+            400,406,422 -> ValidationException()
+            404  -> NotFoundException()
+            502, 503 -> NetworkException()
+            500, 501 -> ServerErrorException()
+            504  -> TimeoutException()
```
Unauthorized, InvalidRequest, AceessDenied..... all these exceptions aren't related to Domain at all. They are remote data source exceptions. 

The only exceptions that domain need are NoInternetConnection and DataSourceException (remote data source and local data source).

mohannadahmed00 replied:
+1

mohamedebrahem13 replied:
there is better way to handle retrofit exception instated of the mapping you can search about retrofit adapter factory

Mostafa-khaled21 replied:
Please check again and tell me if it is right or wrong.

mohamedebrahem13 replied:
Not like this but this is okay for now 👍

Mostafa-khaled21 replied:
Retrofit adapter factory will only deal with the remote exeption. What about the local ?

Mostafa-khaled21 committed:
Refactor: Consolidate network-related exceptions into `NoInternetException`.
- Replacing `NetworkException`, `TimeoutException`, and `ServerErrorException` with a single `NoInternetException`.
- Removing `InvalidRequestMethodException` and `InfrastructureException` as they are now covered by `NoInternetException` or other existing exceptions.
- Updating `SafeCall` to map various data layer exceptions (like `ServerDataException`, `RequestTimeoutDataException`) to the new `NoInternetException`.

Mostafa-khaled21 committed:
Refactor: Remove `UnauthorizedException` and its handling

poula22 reviewed (approved):
Great job 👏👏👏
